### PR TITLE
feat(storage): Add Google Cloud Storage support

### PR DESCRIPTION
**Pull request by kmala**
Add support for storing the backups to GCS.

### CODE REVIEW

File: tests/blackbox.py [hunk 1]
```diff
@@ -46,12 +49,106 @@ class AwsTestConfigSetImpl(AwsTestConfig):
...
-        return AwsTestConfig.patch(self, test_name, default_test_bucket)
+        return AwsTestConfig.patch(self, test_name)
+
+
+class AwsInstanceProfileTestConfig(object):
+    name = 'aws+instance-profile'
+
+    def __init__(self, request):
+        self.request = request
+        self.monkeypatch = request.getfuncargvalue('monkeypatch')
```
Comment (arschles) [inline tests/blackbox.py]: For my info, can you explain what this line is doing?

File: tests/blackbox.py [hunk 2]
```diff
@@ -46,12 +49,106 @@ class AwsTestConfigSetImpl(AwsTestConfig):
...
+            "Code": "Success",
+            "LastUpdated": "3014-01-11T02:13:53Z",
+            "Type": "AWS-HMAC",
+            "AccessKeyId": fed.credentials.access_key,
+            "SecretAccessKey": fed.credentials.secret_key,
+            "Token": fed.credentials.session_token,
+            "Expiration": "3014-01-11T08:16:59Z"
+        }
+
+        from boto import provider
```
Comment (arschles) [inline tests/blackbox.py]: why not put this `import` at the top of the file?

File: tests/blackbox.py (hunk 1 above)
Comment (kmala) [inline tests/blackbox.py]: This code is from a different PR from Wal-e...let me see and update it

File: tests/blackbox.py (hunk 2 above)
Comment (kmala) [inline tests/blackbox.py]: This code is from a different PR from Wal-e....i think we can do that

File: tests/blackbox.py (hunk 1 above)
Comment (bacongobbler) [inline tests/blackbox.py]: @arschles this code is [monkey-patching](http://stackoverflow.com/questions/5626193/what-is-a-monkey-patch) AWS client libs so the integration tests will get an AWS instance profile. Essentially it's mocking up what AWS would typically return from their server so the tests can try to test against the API server. I wouldn't touch it.

Comment (arschles) [inline tests/blackbox.py]: Got it - thanks for the explanation @bacongobbler !

File: tests/blackbox.py (hunk 2 above)
Comment (bacongobbler) [inline tests/blackbox.py]: sometimes if you know that this code will only be used in this one instance, it's best to put it locally to namespace it, so the rest of the code can use the `provider` keyword (which is pretty generic). Again this is best practice, so I wouldn't touch it. The guy knows his stuff ;)

Comment (arschles) [inline tests/blackbox.py]: ah, ok. if it's just moving code around, no need to change it.

**Comment by arschles** [comment]
@kmala there's a lot of added code here that looks like it's from wal-e core. Is that merged from upstream? Anyway, code lgtm overall.

**Comment by bacongobbler** [comment]
This is cherry-picked over from https://github.com/wal-e/wal-e/pull/211, ~~which some users have been saying that they've been running in prod without issues. I'm prone to think that upstream's code is good to :ship:.~~

EDIT: scratch that, looks like there's probably a few tweaks that need to be made. See https://github.com/wal-e/wal-e/pull/211#issuecomment-192332194.

**Comment by kmala** [comment]
@arschles Its from https://github.com/wal-e/wal-e/pull/211 but i had made changes on top of it on how to get connection because the PR assumes we are inside a google cluster and few other things. Also wal-e azure code is broken, so i had to make changes for that also.

**Comment by arschles** [comment]
@bacongobbler @kmala k, sounds like the tweaks are in already then, or am I misunderstanding?

**Comment by bacongobbler** [comment]
I'd probably check to see if there are more relevant fixes to be made with https://github.com/wal-e/wal-e/pull/211#issuecomment-204396723 before testing this one out.

**Comment by kmala** [comment]
@bacongobbler This PR had the changes you mentioned...Can you review it?

**Comment by bacongobbler** [comment]
I don't see the following commits, which were recommended in the original PR:

https://github.com/deverant/wal-e/commit/6b0492c01fd08e823c7ed8f7bc2b5b8ae2e8befb
https://github.com/deverant/wal-e/commit/601dce600b2c76ef7ec616f439fd4d3c48e0faa6
https://github.com/deverant/wal-e/commit/d3bcef1dd2d485fa244cbf73a4ce885ef385e40a

Those should be cherry-picked in as well, especially the first two.

**Comment by bacongobbler** [comment]
Personally I'd hold off until the upstream PR is officially merged so we can maintain upstream compatibility.

**Comment by kmala** [comment]
Like i said the i had made changes to the original PR, we just need the changes of the first commit and that too not all...i had included what is required

**Comment by kmala** [comment]
> Personally I'd hold off until the upstream PR is officially merged so we can maintain upstream compatibility.

Upstream PR is present for 6 months and not much is happening over there and we need these changes for the database to support gcs and azure storage. Also we had already deviated from the upstream on the storage front. Also i had plans to use the object storage cli in place of their storage layer in future.

**Comment by bacongobbler** [comment]
There's been activity as recent as this morning, but if you want this to go in then fine. I have plans at some point in the future to clean up our boto3 crud anyways so hopefully we can get back on the upstream train, so I don't imagine we'll be maintaining this branch forever.

I'll check this out and see how the test runs, since CI isn't hooked up here.

**Comment by bacongobbler** [comment]
I've also found that there are a few issues with this PR when running the test suite:

```
======================================= 10 failed, 122 passed, 64 skipped, 1 pytest-warnings, 23 error in 2.60 seconds =======================================
```

**Comment by kmala** [comment]
I had made changes to the code but forgot to change the test cases....i will be making them now...can you test the flow in the mean time to verify?

File: wal_e/cmd.py [hunk 3]
```diff
@@ -402,19 +424,43 @@ def s3_explicit_creds(args):
     return s3.Credentials(access_key, secret_key, security_token)
 
 
+def s3_instance_profile(args):
+    from wal_e.blobstore import s3
+
+    assert args.aws_instance_profile
+    return s3.InstanceProfileCredentials()
```
Comment (bacongobbler) [inline wal_e/cmd.py]: can you please remove this code? This was removed and no longer exists.

File: wal_e/cmd.py [hunk 4]
```diff
@@ -190,6 +190,23 @@ def build_parser():
                            'the one defined in the programs arguments takes '
                            'precedence.')
 
+    aws_group.add_argument('--aws-instance-profile', action='store_true',
```
Comment (bacongobbler) [inline wal_e/cmd.py]: same here, this option flag doesn't exist

**Comment by bacongobbler** [comment]
can we try to cut this PR into a different branch (like `deis`)? I want to keep the `boto3-upgrade` branch as clean as possible so I can clean that up and merge boto3 support upstream.

**Comment by bacongobbler** [comment]
code LGTM. This branch is only temporary until boto3 support is in wal-e. Are you gonna PR the Azure stuff as well at some point?

**Comment by kmala** [comment]
yes i am planning to upstream the azure stuff.....once all the changes are in wal-e we can shift back to the wal-e

**Pull request merged**